### DISCUSSION TITLE
Offer a combined yield return and yield break syntax

### DISCUSSION THREAD

**Discussion by WillSullivan** [dotnet/csharplang#815] [General]
I've found myself doing the following more than a few times

```
if(someCondition)
{
    yield return Value;
    yield break;
}
```

and often wishing that there was a simplified version, perhaps

```
if(someCondition)
    yield complete Value;
```

I understand the first thing one may think "well, simply place the yield before the check." That does work for simple enumerations, but I've found myself creating more complex enumerations that act as producers and use return values as a way to communicate to consumers.  You might want to yield a work item when processing, but you might want to end after yielding an EOF, or a validation error, or a value indicating cancellation was requested.

The pattern could also be simplified by simply throwing an exception, but these results aren't necessarily exceptional situations, and communicating normal, expected outputs via exceptions is a tad hacky.

**Comment by HaloFour** [comment]
I'm thinking that the need to yield a single result and immediately break the iterator is a pretty uncommon pattern that doesn't necessitate a special syntax.  You're either iterating a value deferred in which case the current syntax is perfectly sufficient, or you're bailing early prior to any deferred calculation in which `yield` probably isn't the best route anyway.

**Reply by SX-GitHub** [reply]
It's one way to reduce "else" to make the code more elegant. Of course the beauty in eye of beholder, many consider "else" statement is "perfect" while others don't. IMO, it is good to have, I prefer "yield return value break;" down below.

**Comment by StefDotmailer** [comment]
Can't this be solved with local functions?

```csharp
private static IEnumerable<int> GetValues(bool someCondition, bool someOtherCondition)
{
    IEnumerable<int> GetValuesLocal()
    {
        foreach (var value in Enumerable.Range(0, 10))
            yield return value;
    }

    if (someCondition)
        return new[] { 1 };

    var valuesLocal = GetValuesLocal();

    if (someOtherCondition)
        return valuesLocal.Concat(new[] { 2 });

    return valuesLocal;
}
```

Although I appreciate you may not want to add the local function structure verbosity, but it will reduce the `yield complete` that you want to a single expression if this is the reason for the request.

And unfortunately the Concat forces you to have an extra allocation.

**Comment by alrz** [comment]
you could invert if to move the yield at the end of method so break is not necessary. in other words, ensure that yield return is the last statement in every code path.

**Comment by ryancdotnet** [comment]
Another idea for the syntax:

```
if (someCondition)
{
    yield return Value break;
}
```

This variant is nearly self-descriptive for anyone who understands the current keywords.

Either way, having such a statement would be nice, at the very least just to remove a superfluous line of code. I'd rather do this than contort a function with if-block-nesting just to get the yield returns at the end.

**Comment by TonyValenti** [comment]
Not a huge fan of this but if accepted, the syntax should be:

yield break x;

To yield x as the final element of a sequence.
